Mathlib/Data/Matroid/Rank/ENat.lean

```diff
@@ -0,0 +1,303 @@
+/-
+Copyright (c) 2025 Peter Nelson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Peter Nelson
+-/
+import Mathlib.Data.ENat.Lattice
+import Mathlib.Data.Matroid.Closure
+
+/-!
+# `ℕ∞`-valued rank
+
+If the 'cardinality' of `s : Set α` is taken to mean the `ℕ∞`-valued term `Set.encard s`,
+then all bases of any `M : Matroid α` have the same cardinality,
+and for each `X : Set α` with `X ⊆ M.E`, all `M`-bases for `X` have the same cardinality.
+The 'rank' of a matroid is the cardinality of all its bases,
+and the 'rank' of a set `X` in a matroid `M` is the cardinality of each `M`-basis of `X`.
+This file defines these two concepts as a term `Matroid.eRank M : ℕ∞`
+and a function `Matroid.eRk M : Set α → ℕ∞` respectively.
+
+The rank function `Matroid.eRk` satisfies three properties, often known as (R1), (R2), (R3):
+* `M.eRk X ≤ Set.encard X`,
+* `M.eRk X ≤ M.eRk Y` for all `X ⊆ Y`,
+* `M.eRk X + M.eRk Y ≥ M.eRk (X ∪ Y) + M.eRk (X ∩ Y)` for all `X, Y`.
```

Comment:
Do you think it makes sense to define a `Prop`-structure `IsRankFunction`, and then use that as an alternative constructor for finite matroids? And also add a proof that `eRk` below is an example...

Maybe this was already discussed somewhere? If so, please link to that discussion.

Reply:
Yes, this makes sense, and something like this is already in the repo. (It was also my fist stab at a definition of matroids in lean3, back in 2020 😃) .

The proofs are a bit of work, so are beyond the current PR, but they are on the long list of contributions that I am still plodding through.

Reply:
The definition and a bunch of related ways to define a matroid from a rank function are here : https://github.com/apnelson1/Matroid/blob/main/Matroid/Axioms/Rank.lean